frame/assets/src/functions.rs

```diff
@@ -571,8 +595,8 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			// Remove source account if it's now dead.
 			if source_account.balance < details.min_balance {
 				debug_assert!(source_account.balance.is_zero(), "checked in prep; qed");
-				if let Remove =
-					Self::dead_account(id, &source, details, &source_account.reason, false)
+				source_died = Remove;
+				if let Remove = Self::dead_account(&source, details, &source_account.reason, false)
```

Comment:
same here, in some case we return that the account has died, and the caller will execute `died` callback. But actually the account can be not dead in case it hold a deposit, and in this case we don't want to execute `died` callback.

Reply:
Yes right, just looking at the return value of `dead_account` is not sufficient.  
The semantic of the `DeadConsequence` seems a bit weird too me now, since even in the `Keep` case, the `died` hook is executed. See [here](https://github.com/paritytech/substrate/blob/8a3434bb5e8f204107644cd32f97f5ba76d8714f/frame/assets/src/functions.rs#L97) where three lines below it is called.  
But it still makes sense, since it is only called when the account dropped below `min-balance`.

Reply:
the `force` case is very specific to when we destroy the whole asset. The usual case is without `force`.
So usually when an account is kept, then FrozenBalance::died is not executed and the account is not removed.

In general I think we should clarify the words here. `dead_account` being executed on account which doesn't die is very weird wording. (but this needs not to be part of your PR)
An account with a deposit held doesn't die. so calling it DeadConsequence is weird to me. It should be called, BelowMinBalanceConsequences.

Reply:
here also I think using a `Option<DeadConsequence>` is good

Reply:
Done